### PR TITLE
[WIP] Major revamp of documentation

### PR DESCRIPTION
**Pull request by Zsailer**
I'm in the middle of adding a bunch of documentation to Jupyter Server. Opening now to get the PR in place for merging.

Taking @kevin-bates's suggestion, I've started by separating the documentation into the following roles/personas—1) Users 2) Operators 3) Developers and 4) Contributors

I'm also planning to add a FAQ section to route readers to the appropriate page based on their questions.

This is rebased on #180. I'm hoping we can merge that soon, then this PR can go in.

### CODE REVIEW

File: docs/source/conf.py [hunk 1]
```diff
@@ -47,12 +47,12 @@
...
+# if os.environ.get('READTHEDOCS', ''):
 
-    # Readthedocs doesn't run our Makefile, so we do this to force it to generate
-    # the config docs.
-    with open('../autogen_config.py') as f:
-        exec(compile(f.read(), '../autogen_config.py', 'exec'), {})
+#     # Readthedocs doesn't run our Makefile, so we do this to force it to generate
+#     # the config docs.
+#     with open('../autogen_config.py') as f:
+#         exec(compile(f.read(), '../autogen_config.py', 'exec'), {})
```
Comment (vidartf) [inline docs/source/conf.py]: Should it stay or should it go? :)

File: docs/source/developers/basic-extension.rst [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+Anatomy of an ``ExtensionApp``
+------------------------------
+
+An ExtensionApp:
+
+    - has traits.
+    - is configurable (from file or CLI)
+    - has a name (see the ``extension_name`` trait).
+    - has an entrypoint, ``jupyter <extension_name>``.
+    - can server static content from the ``/static/<extension_name>/`` endpoint.
```
Comment (kevin-bates) [inline docs/source/developers/basic-extension.rst:110]: typo: server -> serve

File: docs/source/conf.py (hunk 1 above)
Comment (Zsailer) [reply docs/source/conf.py]: Decided to keep this in place. I just temporarily commented it out :)

**Comment by Zsailer** [comment]
I'm done with the docs for now. I'd like to merge this as-is. We can iterate and edit in future PRs.

**Pull request merged**